arch/riscv/core/pmp/core_pmp.c

```diff
@@ -0,0 +1,608 @@
+}
+
+void z_riscv_configure_user_allowed_stack(struct k_thread *thread)
+{
+	unsigned int i;
+
+	z_riscv_pmp_clear_config();
+
+	for (i = 0; i < CONFIG_PMP_SLOT; i++)
+		csr_write_enum(CSR_PMPADDR0 + i, thread->arch.u_pmpaddr[i]);
+
+	for (i = 0; i < RISCV_PMP_CFG_NUM; i++)
+		csr_write_enum(CSR_PMPCFG0 + i, thread->arch.u_pmpcfg[i]);
+}
+
+void z_riscv_pmp_add_dynamic(struct k_thread *thread,
+			ulong_t addr,
+			ulong_t size,
+			unsigned char flags)
+{
+	unsigned char index = 0;
+	unsigned char *uchar_pmpcfg;
+
+	/* Check 4 bytes alignment */
+	__ASSERT(((addr & 0x3) == 0) && ((size & 0x3) == 0) && size,
+		 "address/size are not 4 bytes aligned\n");
+
+	/* Get next free entry */
+	uchar_pmpcfg = (unsigned char *) thread->arch.u_pmpcfg;
+
+	index = PMP_REGION_NUM_FOR_U_THREAD;
+
+	while ((index < CONFIG_PMP_SLOT) && uchar_pmpcfg[index])
+		index++;
+
+	__ASSERT((index < CONFIG_PMP_SLOT), "no free PMP entry\n");
+
+	/* Select the best type */
+	if (size == 4) {
+		thread->arch.u_pmpaddr[index] = addr >> PMP_SHIFT_ADDR;
+		uchar_pmpcfg[index] = flags | PMP_NA4;
+	} else if ((addr & (size - 1)) || (size & (size - 1))) {
+		__ASSERT(((index + 1) < CONFIG_PMP_SLOT),
+			"not enough free PMP entries\n");
+		thread->arch.u_pmpaddr[index] = addr >> PMP_SHIFT_ADDR;
+		uchar_pmpcfg[index++] = flags | PMP_NA4;
+		thread->arch.u_pmpaddr[index] = (addr + size) >> PMP_SHIFT_ADDR;
+		uchar_pmpcfg[index++] = flags | PMP_TOR;
+	} else {
+		thread->arch.u_pmpaddr[index] = (addr >> PMP_SHIFT_ADDR) |
+					((size - 1) >> (PMP_SHIFT_ADDR + 1));
+		uchar_pmpcfg[index] = flags | PMP_NAPOT;
+	}
+}
+
+int arch_buffer_validate(void *addr, size_t size, int write)
+{
+	uint32_t index, i;
+	ulong_t pmp_type, pmp_addr_start, pmp_addr_stop;
+	unsigned char *uchar_pmpcfg;
+	struct k_thread *thread = _current;
+	ulong_t start = (ulong_t) addr;
+	ulong_t access_type = PMP_R;
+	ulong_t napot_mask;
+#ifdef CONFIG_64BIT
+	ulong_t max_bit = 64;
+#else
+	ulong_t max_bit = 32;
+#endif /* CONFIG_64BIT */
+
+	if (write)
+		access_type |= PMP_W;
+
+	uchar_pmpcfg = (unsigned char *) thread->arch.u_pmpcfg;
+
+#ifdef CONFIG_PMP_STACK_GUARD
+	index = 1;
+#else
+	index = 0;
+#endif /* CONFIG_PMP_STACK_GUARD */
+
+__ASSERT((uchar_pmpcfg[index] & PMP_TYPE_MASK) != PMP_TOR,
+	"The 1st PMP entry shouldn't configured as TOR");
+
+	for (; (index < CONFIG_PMP_SLOT) && uchar_pmpcfg[index]; index++) {
+		if ((uchar_pmpcfg[index] & access_type) != access_type)
+			continue;
+
+		pmp_type = uchar_pmpcfg[index] & PMP_TYPE_MASK;
+
+		if (pmp_type == PMP_TOR)
+			continue;
+
+		if (pmp_type == PMP_NA4) {
+			pmp_addr_start =
+				FROM_PMP_ADDR(thread->arch.u_pmpaddr[index]);
+
+			if ((index == CONFIG_PMP_SLOT - 1)  ||
+				((uchar_pmpcfg[index + 1] & PMP_TYPE_MASK)
+					!= PMP_TOR)) {
+				pmp_addr_stop = pmp_addr_start + 4;
+			} else {
+				pmp_addr_stop = FROM_PMP_ADDR(
+					thread->arch.u_pmpaddr[index + 1]);
+				index++;
+			}
+		} else {
+			for (i = 0; i < max_bit; i++) {
```

Comment:
A comment here indicating this branch is handling the NAPOT case would be helpful.

Reply:
Fixed